File: current_session/c++/523_redo.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;

class Solution {
public:
  bool checkSubarraySum(vector<int>& nums, int k) {
    if (nums.size() < 2) return false;
    map<int, int> m;      // use array => MLE
    m[0] = -1;
    int lsum = 0;
    for (int i=0; i<nums.size(); ++i) {
      lsum = (lsum+nums[i])%k;
      if (m.find(lsum) != m.end()) {
        if (i-m[lsum] >= 2) return true;
      } else {
        m[lsum] = i;
      }
    }
    return false;
  }
};

File: current_session/python/523_redo.py
from typing import List

# use dictionary
# we go through the array, and keep adding elements s.t. lsum = sum(A[:i+1])
# key: lsum%k, value: i
# if key exist in dictionary: return true if j-i >= 2
# so on and so forth

class Solution:
    def checkSubarraySum(self, nums: List[int], k: int) -> bool:
        if len(nums) < 2: return False
        lsum = 0
        nd = {0:-1}                 # consider subarray begin from 0
        for j in range(len(nums)):
            lsum = (lsum+nums[j])%k
            if lsum in nd:          # A[i+1:j] is the subarray
                if j-nd[lsum] >= 2: return True
            else:
                nd[lsum] = j
            # print("nd:", nd)
        return False

print(Solution().checkSubarraySum([23,2,4,6,7], 6))
print(Solution().checkSubarraySum([23,2,6,7,4], 6))
print(Solution().checkSubarraySum([23,2,6,4,7], k = 13))
print(Solution().checkSubarraySum([1,2,3],5))

